### PR TITLE
cxx-qt-lib: use C++ templates for common methods

### LINKED ISSUE #376: cxx-qt-lib: Can we use C++ templates to reduce the amount of C++ code
### Discussed in https://github.com/KDAB/cxx-qt/discussions/370

<div type='discussions-op-text'>

<sup>Originally posted by **LeonMatthes** December  6, 2022</sup>
The most common C++ functions we need for cxx-qt-lib are:
- Various constructors (default, copy, etc.)
- Destructor

So could we just add a "cxx-qt-lib/common-utils.h" that defines templates for these.
e.g. for the destructor:
``` cpp
template<class T>
void destruct(T &t) {
    t.~T();
}
```

We've already found that CXX can deal with this pretty well, as long as it can infer all template type arguments.</div>

### PR DESCRIPTION
**Pull request by ahayzen-kdab**
Closes #376

### CODE REVIEW

File: crates/cxx-qt-lib-headers/include/common.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+template<typename T>
+T
+initDefault()
+{
+  return T();
+}
+
+template<typename T>
+T
+clone(const T& value)
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-lib-headers/include/common.h:29]: This is where we could try getting creative:
How about a `construct` method, that can take any argument?
e.g.:
```
template<typename... Args>
T construct(Args.. args) {
    return T(args...);
}
```

Then even stuff like `qDateInit`, `qdatetimeInitFromDateAndTime` could all boil down to this one template

Comment (LeonMatthesKDAB) [reply crates/cxx-qt-lib-headers/include/common.h:29]: We wouldn't even need the `initDefault` method then, just `construct` without arguments would be equivalent.

**Review by LeonMatthesKDAB — CHANGES_REQUESTED** [review]
Going in a great direction. I think we can improve even more here though :thinking:

File: crates/cxx-qt-lib-headers/include/common.h (hunk 1 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-lib-headers/include/common.h]: Done

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

**Comment by Be-ing** [comment]
+277 −724 :+1:

**Pull request merged**